script.artistslideshow/resources/lib/artistslideshow.py

```diff
@@ -0,0 +1,1240 @@
+        self.SLIDESHOWSLEEP = getSettingInt( addon, 'slideshow_sleep', default=1 )
+        self.VALIDIMAGETYPES = tuple( xbmc.getSupportedMedia( 'picture' ).split( '|' )[:-2] )
+        lw.log( ['slideshow thread started'] )
+
+
+    def AddImage( self, path ):
+        if not path:
+            lw.log( ['Image path was empty, nothing added'] )
+            return False
+        if path.endswith( self.VALIDIMAGETYPES ):
+            self.IMAGES.append( path )
+            lw.log( ['Added to image display group: ' + path] )
+            self.IMAGEADDED = True
+            return True
+        else:
+            lw.log( ['Image was not a valid Kodi image type, nothing added: ' + path] )
+            lw.log( ['Valid Kodi image types are:', self.VALIDIMAGETYPES] )
+            return False
+
+
+    def ClearImages( self, fadetoblack ):
+        self.IMAGES = []
+        if fadetoblack:
+            self._set_property( 'ArtistSlideshow.Image', os.path.join( addonpath, 'resources', 'images', 'black-hd.png' ) )
+        else:
+            self._set_property( 'ArtistSlideshow.Image' )
+        lw.log( ['images cleared'] )
+        self.IMAGESCLEARED = True
+
+
+    def run( self ):
+        last_image = ''
+        while self.SHOW:
+            outofimages = True
+            if self._check_for_quit():
+                break
+            if self.IMAGEADDED or self.IMAGESCLEARED or outofimages:
+                random.shuffle( self.IMAGES )
+                self.IMAGEADDED = False
+                self.IMAGESCLEARED = False
+                outofimages = False
+            for image in self.IMAGES:
+                if self.IMAGEADDED or self.IMAGESCLEARED:
+                    lw.log( ['image list changed, resetting loop'] )
+                    break
+                if not image == last_image or len( self.IMAGES ) == 1:
+                    if not image == last_image:
+                        self._set_property( 'ArtistSlideshow.Image', image )
+                        last_image = image
+                    self._wait( wait_time=self.DELAY, sleep_time=self.SLIDESHOWSLEEP )
+                if self._check_for_quit():
+                    break
+        lw.log( ['slideshow thread stopping'] )
+
+
+    def _check_for_quit( self ):
+        cmd = ''
+        with self.QUEUELOCK:
+            if not self.WORKQUEUE.empty():
+                cmd = self.WORKQUEUE.get()
+            if cmd == 'quit':
+                self.SHOW = False
+        return not self.SHOW
+
+
+    def _set_property( self, property_name, value='' ):
+        try:
+          self.WINDOW.setProperty( property_name, value )
+          lw.log( ['%s set to %s' % (property_name, value)] )
+        except Exception as e:
+          lw.log( ['Exception: Could not set property %s to value %s' % (property_name, value), e] )
+
+
+    def _wait( self, wait_time=1, sleep_time=1 ):
+        waited = 0
+        while waited < wait_time:
+            xbmc.sleep( sleep_time*1000 )
```

Comment:
when dealing with seconds please always try to use `xbmc.Monitor().waitForAbort()`. This method  argument can be configured in the settings of your addon so it is highly likely the user doesn't know what he's/she's doing. Kodi will hang on exit if abort is requested while your addon is sleeping in the said conditions.

Reply:
Just seen your comment above. What if you check `xbmc.Monitor().abortRequested()` after the waitForAbort?

Reply:
I'll try that suggestion and see what happens.  I think I might have to do that in both threads to keep AS from hanging Kodi on quit.  I knew you wouldn't like the xbmc.sleep().  '-)

Reply:
It looks like changing the xbmc.sleep() to self.MONITOR.waitForAbort( time ) is OK now.  So maybe it was something else in the code that I cleaned up at some point that was causing the hang.  Interestingly, the slideshow thread will likely never see the abort request because it looks like when Kodi quits it stops the player before sending abort requests to scripts. When the player stops the main thread sends a quit command to the slideshow thread and stops it, so there's a really narrow window that will likely never happen in the real world where an abort request gets to the script before it realizes the player has stopped.